core/src/main/java/dev/morphia/mapping/codec/MorphiaCodecProvider.java

```diff
@@ -4,10 +4,7 @@
 import java.util.HashMap;
```

Comment:
The changes here seem unrelated.  At a minimum it looks like the file was reordered so it makes reading the diff complicated.  Can you confirm if these changes are necessary and if not revert.  If they are necessary can you revert the order back to the original order so the important diffs are easier to review?  Thanks.

Reply:
Sorry, our in-house formatting standard is different. I can revert all the changes, and put them back in using your formatting. But allow me to explain as well...

EntityModel

The EntityModel stores the lifecycle methods in...

    private final Map<Class<? extends Annotation>, List<ClassMethodPair>> lifecycleMethods;

...which ignores the originating class, i.e. the annotated class. So, If I annotate a class A with a @PrePersist, I get an entry in the map from PrepPersist -> ClassMethodPair holding my pre-persist method. That method is now applied across the instances of class A, and all fields of class A which are also entities. This is at a minimum. Maybe the prepersist method is called on classes other than A, or on fields of A that are not entities, I can't tell. The problem is that the call of the prepersist method of A on anything other than instances of A is both incorrect and yields an error, as other types do have that method. My reproducer shows this problem clearly.

This fix I've produced changes the lifecycle map to...

  private final Map<Class<?>, Map<Class<? extends Annotation>, List<ClassMethodPair>>> lifecycleMethods;

...which is to hold a map **for the annotated class**, which points to a map of lifecycle annotations and methods from that class, and which will be applied only to that class in the later processing. All the other changes to EntityModel os in support of that extended structure. That's it.

Now, the code is not good. Neither the original nor the fix. EntityModel is overly complex and weirdly integrated, But the proposed fix is the least intrusive thing I could think of and seems to work in my case. Again, I'd be happier running the tests, can you explain how to do that?

MorphiaCodecProvider

As noted by others the service loading does not work. Fortunately, the codec can be set on MapperOptions. Maybe that was a response to the failure of service locading? Codecs can also be set on ClientSettings. However, we need codecs set by the user to be checked **first**, and there's no way to know who set what codecs in ClientSettings, so I understand the need for another method so the codecs set with it can take priority. The few lines I inserted into MorphiaCodecProvider...

    CodecProvider codecProvider;

    if ((codecProvider = mapper.getOptions().codecProvider()) != null) {
      // Must go first to user defined codecs take precedent
      propertyCodecProviders.add((PropertyCodecProvider)codecProvider);
    }

...are just grabbing the codec provider from MapperOptions, if it exists, and adding it to the codec providers known to MorphiaCodecProvider, at the head of the list, to be checked first. Again, this code all seems to be awkward. From a user perspective the entry point is not great either, Of course, neither was a service provider. an @CodecProvider annotation would be much better in my opinion. But again, given the code, this seems like the least intrusive fix that worked for me.